networkx/algorithms/approximation/traveling_salesman.py

```diff
@@ -0,0 +1,716 @@
+    (due to the presence of the threshold value). In the case of
+    Simulated Annealing, even a very low quality solution can
+    be accepted with probability $p$.
+
+    Time complexity:
+    It has a running time $O(|E||V|^2 + m * n * |V|)$ when a greedy
+    algorithm is used to construct an initial solution, otherwise
+    it has a running time $O(m * n * |V|)$ where $m$ and $n$ are the number
+    of iterations of outer and inner loop respectively.
+
+    For more information and how algorithm is inspired see:
+    http://en.wikipedia.org/wiki/Threshold_accepting
+
+    See Also
+    --------
+    simulated_annealing_tsp
+
+    """
+    if cycle is None:
+        # Construct an initial solution using a greedy algorithm.
+        cycle = greedy_tsp(G, weight=weight, source=source)
+        if G.number_of_nodes() == 2:
+            return cycle
+
+        cost = sum(G[u][v][weight] for u, v in zip(cycle, cycle[1:]))
+    else:
+        # Find the cost of initial solution and make the essential checks for graph.
+        if any(v not in G[u] for u in G for v in G if u != v):
+            raise nx.NetworkXError(
+                "Not a complete graph. All node pairs must have edges."
+            )
+        if not nx.is_weighted(G, weight=weight):
+            raise nx.NetworkXError("Given graph is not weighted.")
+
+        if G.number_of_nodes() == 2:
+            neighbor = list(G.neighbors(source))[0]
+            return [source, neighbor, source]
+
+        cost = sum(G[u][v][weight] for u, v in zip(cycle, cycle[1:]))
+
+    count = 0
+    best_cycle = cycle[:]
+    best_cost = cost
+    while count <= tolerance:
+        count += 1
+        accepted = False
+        for i in range(iterations):
+            adj_sol = _apply_move(cycle, move, seed)
+            adj_cost = sum(G[u][v][weight] for u, v in zip(cycle, cycle[1:]))
+            delta = adj_cost - cost
+            if delta <= threshold:
+                accepted = True
+
+                # Set current solution the adjacent solution.
+                cycle = adj_sol
+                cost = adj_cost
+
+                if cost < best_cost:
+                    count = 0
+                    best_cycle = cycle[:]
+                    best_cost = cost
+        if accepted:
+            threshold -= threshold * alpha
+
+    return best_cycle
+
+
+def _apply_move(soln, move, seed):
+    """
+    Apply a move to a solution to generate a neighbor solution.
+
+    Parameters
+    ----------
+    soln : list of nodes
+        Current solution (list of nodes)
+
+    move : string
+        Move to be applied
+
+    seed : integer, random_state, or None (default)
+        Indicator of random number generation state.
+        See :ref:`Randomness<randomness>`.
+
+    Returns
+    -------
+    The solution after move is applied. (A neighbor solution.)
+    """
+    a = seed.randint(1, len(soln) - 2)
+    listb = list(range(1, a)) + list(range(a + 1, len(soln) - 1))
+    b = seed.choice(listb)
```

Comment:
If I parsed this correctly it looks like the intent here is to sample two indices in the range `[1, len(soln) - 1]` without replacement. If so, then I think the following should work:

```suggestion
    a, b = seed.sample(range(1, len(soln) - 1), k=2)
```

If sampling two indices without replacement wasn't the intent - then maybe add a comment here to explain the intended behavior (it took me a second to parse).

Reply:
Yes, this took me much more than a minute to parse...  
It seems like the chance of the value `len(soln) - 1` is not the same as the chance of the other numbers.

Also, in both the code and your suggestion, 0 is not chosen.  
Why isn't 0 an allowed choice for either a or b?
Maybe it is because the list has the same node at beginning and end of the cycle.
But it could also be that we want to keep the same starting node, or we'll end up chasing our tail around the cycle.  Certainly a comment is needed here.

Reply:
OK... I missed that both 0 and len(soln)-1 are values that this code (either version) does not choose. It's making more sense to me.

It is assuming the list is a cycle and we don't want to switch the first node which must be the same as the last node.  I'll add a comment to this effect.

Reply:
Yeah I wasn't entirely sure on this either - my suggestion based on trying to replicate what the behavior of the original was without truly understanding *why* that was the desired behavior. A comment would definitely be helpful!